### PR TITLE
[ONNX] Enable constant folding for Reshape

### PR DESCRIPTION
**Pull request by neginraoof**
Enabled constant folding for onnx::Reshape

### CODE REVIEW

File: test/onnx/test_utility_funs.py [hunk 1]
```diff
@@ -206,6 +206,22 @@ def forward(self, A):
             assert node.kind() != "onnx::Transpose"
         assert len(list(graph.nodes())) == 1
 
+    def test_constant_fold_reshape(self):
```
Comment (spandantiwari) [inline test/onnx/test_utility_funs.py]: Can we add an end to end test for ORT that matches numbers for a model?

File: test/onnx/test_utility_funs.py [hunk 2]
```diff
@@ -206,6 +206,22 @@ def forward(self, A):
...
+                b = torch.reshape(a, (3, 2))
+                return b + x
+
+        _set_opset_version(self.opset_version)
+        x = torch.ones(3, 2)
+        graph, _, __ = utils._model_to_graph(ReshapeModule(), (x, ), do_constant_folding=True,
+                                             _disable_torch_constant_prop=True)
+        for node in graph.nodes():
+            assert node.kind() != "onnx::Reshape"
+
```
Comment (spandantiwari) [inline test/onnx/test_utility_funs.py]: nit: remove blank line.

File: torch/csrc/jit/passes/onnx/constant_fold.cpp [hunk 3]
```diff
@@ -231,6 +231,15 @@ c10::optional<at::Tensor> runTorchBackendForOnnx(
       return c10::optional<at::Tensor>(updated_val);
     }
     return c10::nullopt;
+  } else if (node->kind() == onnx::Reshape) {
+    assert(inputTensorValues.size() == 2);
+    updated_val = inputTensorValues[0];
+    std::vector<int64_t> shape(inputTensorValues[1].sizes()[0], 0);
```
Comment (spandantiwari) [inline torch/csrc/jit/passes/onnx/constant_fold.cpp]: Consider adding a check if the number of elements between old and new shape match.

File: torch/csrc/jit/passes/onnx/constant_fold.cpp [hunk 4]
```diff
@@ -231,6 +231,15 @@ c10::optional<at::Tensor> runTorchBackendForOnnx(
...
+    auto shape_a = inputTensorValues[1].accessor<int64_t, 1>();
+    for (size_t i = 0; i < inputTensorValues[1].sizes()[0]; ++i) {
+      shape[i] = shape_a[i];
```
Comment (spandantiwari) [inline torch/csrc/jit/passes/onnx/constant_fold.cpp]: Since we are dealing with an ONNX graph, it is possible that the reshape node's second input has `-1` or `0` in the new shape. The code here needs to handle those situations, before passing it along to ATen op.

File: test/onnx/test_utility_funs.py (hunk 1 above)
Comment (neginraoof) [reply test/onnx/test_utility_funs.py]: Sure. I was considering the mask_rcnn test for that that I'm adding in a separate PR with other torchvision tests. I can include that test in this PR.

File: torch/csrc/jit/passes/onnx/constant_fold.cpp (hunk 4 above)
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/constant_fold.cpp]: Thanks for pointing this out.
-1 dim value is actually supported in aten. This is actually tested with mask rcnn model.
But for 0 values, I updated the PR to address this.

File: torch/csrc/jit/passes/onnx/constant_fold.cpp (hunk 3 above)
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/constant_fold.cpp]: The new shape array is defined to have the same size as the initial shape input
std::vector<int64_t> shape(inputTensorValues[1].sizes()[0], 0);
And when filling the shape array, we iterate over the initial shape input
for (size_t i = 0; i < inputTensorValues[1].sizes()[0]; ++i)

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit 076a343:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=076a343c5be5f399eaeb99fca5ca015f1a7d05de).

### 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/3911680) caffe2_onnx_py3_6_clang7_ubuntu16_04_test (1/1)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=72715637))
```
Dec 12 01:37:25 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_index_put_ellipsis SKIPPED [ 94%]
Dec 12 01:37:25 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_index_put_slice_index SKIPPED [ 94%]
Dec 12 01:37:25 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_index_select_constant_scaler_index PASSED [ 94%]
Dec 12 01:37:25 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_index_select_scaler_index PASSED [ 94%]
Dec 12 01:37:25 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_inplace_arithmetic PASSED [ 94%]
Dec 12 01:37:25 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_inplace_fill PASSED [ 94%]
Dec 12 01:37:25 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_inplace_zero PASSED [ 94%]
Dec 12 01:37:25 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_interpolate_downsample SKIPPED [ 94%]
Dec 12 01:37:25 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_interpolate_no_shape SKIPPED [ 94%]
Dec 12 01:37:26 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_interpolate_upsample PASSED [ 94%]
Dec 12 01:37:26 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_keypoint_rcnn FAILED [ 94%]
Dec 12 01:37:26 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_layer_norm PASSED [ 94%]
Dec 12 01:37:26 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_log PASSED [ 94%]
Dec 12 01:37:26 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_log1p PASSED [ 94%]
Dec 12 01:37:26 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_logdet SKIPPED [ 94%]
Dec 12 01:37:26 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_logsoftmax PASSED [ 94%]
Dec 12 01:37:26 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_logsoftmax_dim PASSED [ 94%]
Dec 12 01:37:26 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_lstm PASSED [ 94%]
Dec 12 01:37:27 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_lstm_constant_folding PASSED [ 94%]
Dec 12 01:37:27 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_lstm_default_init_state PASSED [ 94%]
Dec 12 01:37:27 test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_lstm_fixed_batch_size PASSED [ 94%]
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

**Comment by neginraoof** [comment]
The failure is caused due to the missing symbolic for the recently added floor_divide op. This op is supported in https://github.com/pytorch/pytorch/pull/31081
Once #31081 is merged, tests should pass

File: torch/csrc/jit/passes/onnx/constant_fold.cpp (hunk 3 above)
Comment (spandantiwari) [reply torch/csrc/jit/passes/onnx/constant_fold.cpp]: That's fine. But consider a situation where a bad input shape comes in which has number of elements different than the original input array. That could crash the program in the loop below, most probably. I was suggesting having a check that would validate that the two inputs agree on the shape so that we can error out gracefully if they don't.

File: torch/csrc/jit/passes/onnx/constant_fold.cpp [hunk 5]
```diff
@@ -231,6 +231,26 @@ c10::optional<at::Tensor> runTorchBackendForOnnx(
...
+    std::vector<int64_t> shape(inputTensorValues[1].sizes()[0], 0);
+    auto shape_a = inputTensorValues[1].accessor<int64_t, 1>();
+    for (size_t i = 0; i < inputTensorValues[1].sizes()[0]; ++i) {
+      // All shape dim values should be >= -1
+      // onnx::Reshape supports a shape dim value to be zero, in
+      // which case the actual dim value remains unchanged. However,
+      // at::reshape does not support shape dim value to be zero
+      assert(shape_a[i] >= -1);
+      if (shape_a[i] == 0){
+        assert(i < inputTensorValues[0].sizes().size());  // dim with value zero < the dimension size input tensor
```
Comment (spandantiwari) [inline torch/csrc/jit/passes/onnx/constant_fold.cpp]: This is a good check. In this particular instance we should have an explicit error message spelling out the issue instead of an assert.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 6]
```diff
@@ -2274,6 +2337,7 @@ def test_constant_pad(self):
         x = torch.randn(2, 2, 4, 4)
         self.run_test(model, x)
 
+
```
Comment (spandantiwari) [inline test/onnx/test_pytorch_onnx_onnxruntime.py]: nit: extra line.

**Review by spandantiwari — APPROVED** [review]
LGTM.

**Comment by neginraoof** [comment]
Failure in TestAllCompare.test_allcompare. Looks like this is unrelated.

**Review by houseroad — COMMENTED** [review]
ONNX test has failure, could you take a look and fix it?

**Comment by neginraoof** [comment]
> ONNX test has failure, could you take a look and fix it?

ONNX related tests are passing now. I think CI failures are unrelated.
